### PR TITLE
Read API part 1: C++ API query and buffer management libraries

### PR DESCRIPTION
**Pull request by ihnorton**
This commit introduces two libraries for use by the high-level API:

  - BufferSet: this class encapsulates the data and (conditionally)
    offsets and validity buffers to return results of a TileDB query.
  - ManagedQuery: this class handles running a TileDB query to completion
    and returning results based on a set of point or range selections.
    It is designed to handle bookkeeping around TileDB input
    buffers and potential batch mode usage (incomplete queries),
    in order to simplify the high-level API interface.

### CODE REVIEW

**Comment by ihnorton** [comment]
The key part of this PR is this test section:

https://github.com/single-cell-data/TileDB-SingleCell/pull/3/files#diff-be772ca0709353d89ca9078269d59253170a4c820d5a27410f097a54e1f700a1R43-R45

File: libtiledbsc/src/lib/buffer_set.cc [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+BufferSet::BufferSet(
+    std::string name,
+    size_t nelem, size_t elem_nbytes,
+    bool isvar, bool isnullable)
+    :
+    name_(name), elem_nbytes_(elem_nbytes), offsets(std::nullopt), validity(std::nullopt)
+{
+    data = std::vector<DELEM_T>(nelem * elem_nbytes);
+    //std::cout << "0a name: " << name << " data: " << (void*)data.data() << std::endl;
```
Comment (Shelnutt2) [inline libtiledbsc/src/lib/buffer_set.cc:34]: Left over?

File: libtiledbsc/src/lib/buffer_set.cc [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    if (isvar) {
+        //offsets = make_optional(alloc(nelem_offsets_nbytes(nelem)));
+        offsets = std::vector<DELEM_T>(nelem_offsets_nbytes(nelem));
+        //std::cout << "0b name: " << name << " data: " << offsets.value().data() << std::endl;
```
Comment (Shelnutt2) [inline libtiledbsc/src/lib/buffer_set.cc:38]: Left over?

File: libtiledbsc/src/lib/buffer_set.cc [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    if (isnullable) {
+        //validity = make_optional(alloc(nelem_validity_nbytes(nelem)));
```
Comment (Shelnutt2) [inline libtiledbsc/src/lib/buffer_set.cc:41]: Leftover?

**Review by Shelnutt2 — APPROVED** [review]
Minor nits

**Pull request merged**